src/corehost/common/pal.windows.cpp

```diff
@@ -76,6 +76,46 @@ bool pal::touch_file(const pal::string_t& path)
     return true;
 }
 
+void* pal::map_file_readonly(const pal::string_t& path, size_t &length)
+{
+	HANDLE file = CreateFileW(path.c_str(), GENERIC_READ, FILE_SHARE_READ, NULL, OPEN_EXISTING, FILE_ATTRIBUTE_NORMAL, NULL);
+
+	if (file == INVALID_HANDLE_VALUE)
+	{
+		trace::warning(_X("Failed to map file. CreateFileW(%s) failed with error %d"), path.c_str(), GetLastError());
+		return nullptr;
+	}
+
+	LARGE_INTEGER fileSize;
+	if (GetFileSizeEx(file, &fileSize) == 0)
+	{
+		trace::warning(_X("Failed to map file. GetFileSizeEx(%s) failed with error %d"), path.c_str(), GetLastError());
+        CloseHandle(file);
+		return nullptr;
+	}
+	length = (size_t)fileSize.QuadPart;
+
+	HANDLE map = CreateFileMappingW(file, NULL, PAGE_READONLY, 0, 0, NULL);
+
+	if (map == NULL)
+	{
+		trace::warning(_X("Failed to map file. CreateFileMappingW(%s) failed with error %d"), path.c_str(), GetLastError());
+		CloseHandle(file);
+		return nullptr;
+	}
+
+	void *address = MapViewOfFile(map, FILE_MAP_READ, 0, 0, 0);
+
+	if (map == NULL)
+	{
+		trace::warning(_X("Failed to map file. MapViewOfFile(%s) failed with error %d"), path.c_str(), GetLastError());
+		CloseHandle(file);
+		return nullptr;
+	}
+
+	return address;
```

Comment:
I think the function needs a comment about the fact that there's no way to close the mapping (or the file).
I wonder if this is actually OK for the bundler. Especially in this stage where everything is extracted to disk - we would be keeping potentially large amount of memory blocked by the file mapping. And we would unnecessary stress on the memory manager by using all those pages (to extract the files to disk), but never freeing those. Eventually they would get evicted from physical ram, but it feel unnecessary.
We should either close the mapping (and thus map part by part probably), or if there's a way tell the memory managed to discard the mapped pages we know we're not going to need anymore.

Reply:
We keep the mapping file open, because it's necessary to access parts of the file later (tbd in further stages). 

I don't understand the concern here.  Is it:
* Running out of virtual addresses? 
    * Even on 32 bit architectures, if the app is so large that it is running short of virtual addresses, I it is not a good candidate to be published as a single-exe. 
    * We can unmap the file after extraction to disk now, but such large apps will start to fail once we try to load files directly from bundle. 
    * Alternatively, we can map parts of the file (which is tantamount to the existing `fread-fwrite` solution). But I don't think we should optimize for this case.
* Paging ? 
    * The bundle is mapped as `O_RDONLY`. So, there is no overhead to evicting the pages -- they're just gone when physical memory is needed. 
    * I don't see why these pages should be released upfront, that should handled by the OS as needed.

Reply:
The concern is both. I'm not saying that either is blocking, but I would like us to consider improving this. The runtime (and this is part of the runtime effectively) always tries to be as efficient as possible to give most resources to the app itself.

VM space usage is something which some apps may be sensitive to, if nothing else, people often just look at the memory usage of an app and complain if it's too high. I agree that mapping each file separately is probably too much of a hassle (but it should still be better than the current read/write). That said when we're creating the bundle we know pretty well which files we're going to extract and which will be potentially read from memory. We could order the files in the bundle into two groups and then memory map the MSIL as a whole. For the files which will be spilled to disk I would probably go with memory mapping per file.

Paging - I agree that the OS will have no problem evicting the pages, the point was that for the files which were extracted we know for a fact that the memory will never be accessed again. It would be a very nice to tell OS about that fact. Otherwise OS will treat these pages the same as any other RO pages in the process and potentially chose RO pages which will be needed again for eviction over those we know will never be used.

Both these points are relatively subtle, but both are something we typically tend to look into as part of being "in the runtime".

Reply:
I think we should unmap the bundle file in the following cases:
* When the apphost is not a bundle
* When the apphost runs in self-extractor mode
    * Currently always
    * In later stages, when [`ExtractAllFiles`](https://github.com/dotnet/designs/blob/master/accepted/single-file/extract.md#extraction-configurations) property is set. 

I'll make this change. 

However, when files are actually used directly from the bundle, I'm not sure if mapping/unmapping in parts is warranted. I say this because:

* At startup, only the pages that are actually extracted out to disk will be touched, the rest of the file will only take up virtual-addresses, not the memory backing those addresses. So, this will not count towards memory footprint of the app.
* Mapping/unmapping in parts causes additional system call overhead during startup.
* As you mentioned, we could consider splitting the bundle into two parts -- (i) Files to be extracted, and (ii) Files to be used directly from bundle, plus meta-data -- and load them separately. But as implementation progresses to further stages, most of the files will be in the second set that lives through the process. 
* Regarding paging, I suggest that we measure the impact (if any) on some large apps once assemblies are actually loaded from the bundle. We can take mitigating steps if necessary.